src/daft-logical-plan/src/optimization/rules/granular_projections.rs

```diff
@@ -0,0 +1,389 @@
+/// URL downloads, but this may be extended in the future to other functions and Python UDFs.
+#[derive(Debug)]
+pub struct GranularProjections {}
+
+impl OptimizerRule for GranularProjections {
+    fn try_optimize(&self, plan: Arc<LogicalPlan>) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        plan.transform_up(|node| self.try_optimize_node(node))
+    }
+}
+
+impl GranularProjections {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn requires_granular_morsel_sizing(expr: &Expr) -> bool {
+        // TODO: Add Python UDFs as well, but need to handle multiple args better
+        // As well as good testing
+        matches!(
+            expr,
+            Expr::ScalarFunction(ScalarFunction { udf, .. }) if udf.name() == "url_download"
+        )
+    }
+
+    fn try_optimize_node(
+        &self,
+        plan: Arc<LogicalPlan>,
+    ) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        // Only apply to Project nodes, skip others
+        let LogicalPlan::Project(Project {
+            projection, input, ..
+        }) = plan.as_ref()
+        else {
+            return Ok(Transformed::no(plan));
+        };
+
+        // Contains all of expression pieces for each projection column
+        let mut all_split_exprs = vec![];
+
+        // Check if the projection has operations to isolate
+        for expr in projection {
+            let mut split_exprs = vec![];
+
+            let res = expr.clone().transform_up(|e| {
+                // If we encounter a function that needs to be isolated, we split off its children
+                // and replace with column references.
+
+                // TODO: Should be have a helper method or something to identify these
+                // operations that need to be treated specially. Esp for future?
+                if Self::requires_granular_morsel_sizing(e.as_ref()) {
+                    // Split and save child expressions
+                    let mut children = vec![];
+                    let mut changed = false;
+                    for child in e.children() {
+                        match child.as_ref() {
+                            // If child is just a reference, don't do anything
+                            Expr::Column(_) | Expr::Literal(_) => children.push(child),
+                            _ => {
+                                changed = true;
+                                // Child may not have an alias, so we need to generate a new one
+                                // TODO: Remove with ordinals
+                                let child_name = format!("id-{}", uuid::Uuid::new_v4());
+                                let child = child.alias(child_name.as_str());
+                                split_exprs.push(child);
+
+                                children.push(Arc::new(Expr::Column(Column::Resolved(
+                                    ResolvedColumn::Basic(child_name.into()),
+                                ))));
```

Comment:
Small nit: we do have a `resolved_col` method that allows you to easily construct these.